Libertydevops 1 gitjenkinshelmiks (#2245)
OpenLiberty/blogs: posts/2022-04-15-libertydevops-gitjenkinshelmiks.adoc

                         docker build -t liberty-$<code identifier>:$<docker image version> .
                         #eg: docker build -t liberty-app:v1.0 .
                         docker tag liberty-$<code identifier>:$<docker image version> <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker tag liberty-app:v1.0 docker.io/someid/liberty-app:v1.0
                         docker push <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker push docker.io/someid/liberty-app:v1.0  
                       '''   
                  }
               }
            }
          }
----

Where:

* _git-token_ is the personal access token you have generated from your Github account.
* _remote-docker-image-repository-url_ is the location of the Docker image repository.
* _username_ is your user name for the Docker image repository.
* _password_ is your password for the Docker image repository.
* _docker-image-version_ is the version number of your Docker image, a unique identifier.

> Use this link:/guides/containerize.html[sample Dockerfile] and customize it for your application.

Reviewer: Consider the following:

"Build your container image using a Dockerfile or a Containerfile generated by Liberty Starter ([start.openliberty.io](http://start.openliberty.io/)) or by following the pattern described at https://github.com/OpenLiberty/ci.docker.   The Containerize guide (https://openliberty.io/guides/containerize.html) is a helpful resource that goes into more details on how to create a container image for applications running on Liberty."

Cc @lauracowen

Author: Done

Apart from running JUnit test cases along with the code build phase, Jenkins can trigger functional and integration QA test cases automatically after deploying the cloud-native Java application.

Configure the test cases in the Jenkins job and test it manually. Create a remote job identifier authentication token in the "Trigger builds remotely" section under "Build Triggers". Trigger this test case from the Docker "entrypoint" file by using a remote rest API call that uses this authentication token as the identifier.

For example, run the following command in a terminal: 

[source]
----
curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<remote-job-identifier-authentication-token>
----

You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the Jenkins login credentials.

== Kubernetes monitoring tools

Kubernetes provides commands to check the application or cluster logs and memory and CPU usage through the commands like: 

[source]
----
kubectl logs ..
cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)
----

> There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. 

Reviewer: Let us make a reference to Liberty's support for these tools too?

"Liberty makes it easy for system and application metrics to be collected for observability.  See [additional resources](https://community.ibm.com/community/user/wasdevops/blogs/don-bourne1/2021/06/26/metrics-and-monitoring-guidance-for-open-liberty-a) on integration with Prometheus and Grafana."

Author: Done

---
layout: post
title: "Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins"
categories: blog
author_picture: https://avatars3.githubusercontent.com/shamjithantholi
author_github: https://github.com/shamjithantholi

> seo-title: TITLE - OpenLiberty.io

Reviewer 2: Replace TITLE with the "Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins".

Author: Done

> seo-description: DESCRIPTION

Reviewer 2: Replace DESCRIPTION with "How to build a DevOps pipeline for a cloud-native Java application with Jenkins as the core controller component."
(without the quotation marks for this one)

Author: Done

> blog_description: "A high level explanation about how to build a devops pipeline with Jenkins as the core controller component "

Reviewer 2: Change "A high level explanation about how to build a devops pipeline with Jenkins as the core controller component" to "How to build a DevOps pipeline for a cloud-native Java application with Jenkins as the core controller component."

Author: Done

> open-graph-image: https://openliberty.io/img/twitter_card.jpg

Reviewer 2: Could you replace the URL here with:

https://openliberty.io/img/blog/liberty-devops-generic-architecture-2.png

(it just means that your nice architecture diagram is automatically used in social media posts instead of the generic OL logo)

Author: Done

:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then automatically starts a code build, security scanning, and test running, and deployment of the application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[OpenLiberty].

Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. For example, Jenkins can pull code changes from a git repository, build the code using Maven, then run Junit tests against the built code, and deploy the application into Kubernetes.

> The following diagram shows a simple architecture for a pipeline that builds and deploys a Jakarta EE & MicroProfile application in a single Kubernetes environment. When a code change is detected by Jenkins, Jenkins builds the code, runs security and static analysis scans, and runs unit tests against the built code. Jenkins then generates a Docker image of the application with its Open Liberty server and saves the image to a Docker repository, Jenkins then trigger the deployment using CLI commands and deploy the Docker image to Kubernetes..

Reviewer 2: Typos in the last line:
"Jenkins then trigger the deployment using CLI commands and deploy the Docker image to Kubernetes.." should be "Jenkins then triggers the deployment using CLI commands and deploys the Docker image to Kubernetes."

Author: Done

> image::/img/blog/liberty-devops-generic-architecture-2.png[Liberty DevOps architecture diagram ,width=70%,align="center"]

Reviewer 2: For all images in the body of the post, can you talk to @kinueng about whether we still have a style tag that sets a very fine, light grey border around the images, please? It makes the edges look more defined and less scrappy. It's more useful for screenshots than for diagrams but see how it looks.

> I will assume that you have a basic understanding of git, Docker, Kubernetes. In this blog post, I will assume that the Jakarta EE & MicroProfile application code are stored in link:https://github.com/[Github]. I will use link:https://hub.docker.com/[DockerHub] to store the containerized application (correction: Docker images)and I will deploy the containerized application to link:https://cloud.ibm.com/kubernetes/catalog/create[IBM Cloud Kubernetes Service].

Reviewer 2: I think we should move "In this blog post," to the start of the paragraph, so that it reads "In this blog post, I will assume that you have a basic understanding...." then the next sentence starts "I will also assume...".

Change "code are stored" to "code is stored"

Author: Done

> == Installing and configuring Jenkins to setup CI/CD of Jakarta EE & MicroProfile application

Reviewer 2: Change to:
"Installing and configuring Jenkins to set up CI/CD of a cloud-native Java application"

Author: Done

> 

Reviewer 2: Insert here: "You can install Jenkins on any physical or virtual server or it can run as a container on Kubernetes itself."

Then remove that sentence from the next paragraph.

Author: Added with "Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. For example, Jenkins can pull code changes from a git repository, build the code using Maven, then run Junit tests against the built code, and deploy the application into Kubernetes. You can install Jenkins on any physical or virtual servers, or it can be running as a container on Kubernetes itself."

> Install *jenkins* with all the recommended plug-ins link:https://www.jenkins.io/doc/book/installing/[install Jenkins], make sure that the following plug-ins are installed. You can install Jenkins on any physical or virtual servers, or it can be running as a container on Kubernetes itself.

Reviewer 2: Remove the bold highlighting from "jenkins" and capitalise Jenkins.

Author: Done

> Install *jenkins* with all the recommended plug-ins link:https://www.jenkins.io/doc/book/installing/[install Jenkins], make sure that the following plug-ins are installed. You can install Jenkins on any physical or virtual servers, or it can be running as a container on Kubernetes itself.

Reviewer 2: Remove "You can install Jenkins on any...." sentence.

Author: Done

> Install *jenkins* with all the recommended plug-ins link:https://www.jenkins.io/doc/book/installing/[install Jenkins], make sure that the following plug-ins are installed. You can install Jenkins on any physical or virtual servers, or it can be running as a container on Kubernetes itself.

Reviewer 2: Reword to "Install link:[https://www.jenkins.io/doc/book/installing/[Jenkins]](https://www.jenkins.io/doc/book/installing/%5BJenkins%5D) with the following plugins on the base image of your Jenkins container image:"

Author: Replaced with "Install link:https://www.jenkins.io/doc/book/installing/[Jenkins] with the following plugins on the base image of your Jenkins container image, or on the Jenkins master."

> 

Reviewer 2: For each of the following plugins, give a brief description of why to install it (replace "xxx" with the actual purpose):

"* Maven, to build the Java code
* Pipeline, to xxxx 
* Multibranch Scan Webhook Trigger, to xxx
* Docker, to create a container image
* Kubernetes, to xxx"

Author: Done

* Maven
* Pipeline
* Multibranch Scan Webhook Trigger
* Docker
* Kubernetes

> Jenkins builds the Jakarta EE & MicroProfile application code by using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins slave agent (good configuration server (or container)). More details about Jenkins slave setup are provided at link:https://www.jenkins.io/doc/book/using/using-agents/[Jenkins slave setup], link:https://www.jenkins.io/doc/book/pipeline/syntax/[pipeline code details] 

Reviewer 2: Replace this paragraph with (note the changed terminology from "slave" to "agent" - this matches what Jenkins uses on the pages linked to):

"Jenkins builds the Liberty Java code using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:[https://www.jenkins.io/doc/book/using/using-agents/](https://www.jenkins.io/doc/book/using/using-agents/%3E)[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/[Pipeline](https://www.jenkins.io/doc/book/pipeline/syntax/%5BPipeline) Syntax]."

Author: Done

> == Jenkins pipeline scripting introduction

Reviewer 2: Change to "== Writing Jenkins pipeline scripts" (or similar - make it a bit more active, and match the style of the other headings too)

Author: Done

> It is recommended to adhere to the concept of infrastructure as a code (IaaC) in DevOps pipeline, pipeline scripting-based jenkins jobs are a good example of IaaC.

Reviewer 2: Re-word slightly to "It is good practice to adhere to the concept of infrastructure as a code (IaaC) when writing DevOps pipelines. Writing Jenkins jobs as pipeline scripts is a good example of IaaC."

Author: Done

You can write link:https://www.jenkins.io/doc/book/pipeline/syntax/[Jenkins pipeline code] in one of the following ways:

* Directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI]. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 

> image::/img/blog/pipeline-code-on-jenkins.png[Pipeline code directly on Jenkins ,width=70%,align="center"]

Reviewer 2: Slightly rephrase the alt text to "Pipeline code written directly in the Jenkins UI and stored in the Jenkins installation."

Author: written "Pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins master"

> * Jenkinsfile (a plain text file) in Git and mapping it to Jenkins. This is better for ensuring that your configuration is always under version control. If you have specific build and deployment configurations for separate environments, such as dev, staging, and production, you create a separate Jenkinsfile for each environment and store it in that environment specific Git repo  branches. You can either use "Pipeline" type job or "Multibranch pipeline" type job in this case.

Reviewer 2: Change the start to "In a Jenkinsfile (a plain text file)..." to match the previous bullet point in style.

Reviewer 2: Change "environment specific Git repo branches" to "environment-specific git repository branches"

Reviewer 2: Rephrase slightly from "You can either use "Pipeline" type job or "Multibranch pipeline" type job in this case." to "You can use either the "Pipeline" or "Multibranch pipeline" type of job in this case."

Author: Done

> image::/img/blog/pipeline-code-on-git.png[Pipeline code on Git ,width=50%,align="center"]

Reviewer 2: Rephrase the alt text to "Pipeline code stored in a Jenkinsfile on git."

Author: Done

> == Building the Jakarta EE & MicroProfile application with Open Liberty on Jenkins

Reviewer 2: "== Building the cloud-native Java application with Open Liberty on Jenkins" (for consistency/SEO with the previous headings).

Author: Done

> Before writing pipeline script to automate the deployment, you need to make a change in Dockerfile. The standard Dockerfile that you use in general may not enough to do the Jakarta EE & MicroProfile application deployment (to OpenLiberty server) on containerized environment. A sample Openliberty compliant Dockerfile snipped in given, you can configure it as needed (gather the additional files mentioned in this code from open liberty support team).

Reviewer 2: What's changed in this sample? It's not clear from this explanation what needs changing. Couldn't we keep it simple here and just say to use this (greatly simplified) Dockerfile?

(the text in this explanation needs tidying up for grammar and capitalisation but I'm not sure what it needs to say so I'll not do that just now)

Author: Removed everything and kept only "Build your Dockerfile from the foundation code given below (gather the additional files mentioned in this code from open liberty support team)."

    FROM icr.io/appcafe/open-liberty:full-java11-openj9-ubi
    # Add Liberty server configuration including all necessary features
    COPY --chown=1001:0  server.xml /config/
    # Modify feature repository (optional)
    COPY --chown=1001:0 featureUtility.properties /opt/ol/wlp/etc/
    # This script will add the requested XML snippets to enable Liberty features and grow image to be fit-for-purpose using featureUtility. 
    # Only available in 'kernel-slim'. The 'full' tag already includes all features for convenience.
    RUN features.sh
    # Add interim fixes (optional)
    #COPY --chown=1001:0  interim-fixes /opt/ol/fixes/
    .
    .
    .
    .
    COPY --chown=1001:0 src/main/liberty/config/ /config/
    COPY --chown=1001:0 ./target/*.war /config/apps/
    RUN configure.sh

> Below given sample pipeline code can perform the code build, packaging and generating docker image and pushing the same to remote docker hub. You can use it by modifying the parameter section (<>)

Reviewer 2: Can you change the style of this paragraph, the code, and the parameter explanations to this:

"You configure a Jenkins job by adding build actions. The following sample pipeline code builds the Liberty Java code, packages it into a Docker image, and pushes the Docker image to a remote Docker image repository, such as DockerHub:"

Put the code snippet here. Alter the parameter style/format to match the list below.

"Where:

* git-token is the personal access token you have generated from your Github account.
* remote-docker-image-repository-url is the location of the Docker image repository.
* username is your user name for the Docker image repository.
* password is your password for the Docker image repository.
* docker-image-version is the version number of your Docker image, a unique identifier."

Don't include the optional parameters here. They can find out more in the Docker docs.

Author: Avoiding the line "You configure a Jenkins job by adding build actions." because no "Build actions" in pipeline type job

Author: Other changes are done

                         docker tag liberty-$<code identifier>:$<docker image version> <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker tag liberty-app:v1.0 docker.io/someid/liberty-app:v1.0
                         docker push <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker push docker.io/someid/liberty-app:v1.0  
                       '''   
                  }
               }
            }
          }

Following are the parameter that is used in this example code

* Git token: Personal access token generated from github.
* remote-docker-image-repository-url : Docker hub repository URL.
* USERNAME/PASSWWORD: username and password to connect to docker registry.

image::/img/blog/jenkins-cred-binding-and-corresponding-param.png[Pipeline credential binding and corresponding param,width=30%,align="center"]

* code identifier: This is optional, a unique docker image identifier
* docker image version: docker image version number, a unique identifier   

> *Security scan* can be done along with Maven build command (CLI commands way of application scanning) or can do it in a separate pipeline stage. 

Reviewer 2: Rephrase to: "You can add build actions to perform static analysis of your code and security scans. For example, you can use link:[https://www.sonarqube.org/downloads/[SonarQube](https://www.sonarqube.org/downloads/%5BSonarQube) community edition] to do static code analysis, and link:[https://hub.docker.com/r/aquasec/trivy/[trivy]](https://hub.docker.com/r/aquasec/trivy/%5Btrivy%5D) to do a security scan of your built Docker image."

Author: This is not a correct way of explaining security scan setup and actual scan. For sonarqube scan, we need to setup scan server separately and then the scan client commands  which connects with the server for performing scan analysis. If I replace the whole explanation about scan setup as you suggested, I worry that users will be confused about the whole setup. We can discuss this further

> For static code analysis, SonarQube community edition can be used. Install SonarQube server by either using file start-up type from cli by downloading the package in link:https://www.sonarqube.org/success-download-community-edition/[SonarQube server installation package] or use docker way as explained in link:https://docs.sonarqube.org/latest/setup/get-started-2-minutes/[SonarQube server installation steps]. SonarQube jenkins client setup details are given at link:https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-jenkins/[SonarQube client for jenkins]. 

Reviewer 2: Remove this paragraph - too much detail.

 
    CLI command example, integrated with Maven command: 
    $ mvn package verify sonar:sonar -Dsonar.projectKey=sampleapp -Dsonar.host.url=http://localhost:9000 -Dsonar.login=7b13f240123456780479b13d09ba8
   
For Docker image scan, you can use link:https://hub.docker.com/r/aquasec/trivy/[Docker image scan with trivy]. This scan provides the vulnerability details of open source jars that are used in the application. Sample commands are given below

       sh '''
             docker run aquasec/trivy image $docker_login/sampleapp:v1.0
             #eg: docker run aquasec/trivy image docker.io/someid/liberty-app:v1.0
       '''

> 

Reviewer 2: Remove this detail. It's too much detail that they can find in the respective docs.

Author: There is actually very less details in the scan setup section, there are a lot more to explain in general, how to interpret scan results regarding code vulnerabilities and coverage percentage, how to push selected scan results to server, how to fail the build job in case of not satisfying coverage percentage, how to develop quality gate setup, where can we see the test case execution details in case of failures, why is the coverage percentage zero in case of failures etc etc. its a long list and nothing is explained in this section. Actually none of these things will be available on official documents directly)  . I have removed the line "*Security scan* can be done along with Maven build command (CLI commands way of application scanning) or can do it in a separate pipeline stage. ", other lines are very important for the clarity. Steps followed are like this. First, explained the SonarQube server setup, on both VM and containerized environment. Then the client setup, an example is given after that following the pattern in other sections. We can discuss this further.

> == Deploying Jakarta EE & MicroProfile applications with Open Liberty to Kubernetes with Jenkins

Reviewer 2: "== Deploying cloud-native Java applications with Open Liberty to Kubernetes with Jenkins"

Author: Done

> For simplicity, I will use the command line (CLI) option to configure Jenkins to deploy a Jakarta EE & MicroProfile application with Open Liberty to Kubernetes. You could use one of many tools though, such as Helm, Travis CI, Circle CI, etc.

Reviewer 2: change "Jakarta EE & MicroProfile application" to "cloud-native Java" application".

I appreciate that we might need to say that somewhere but I think get it all consistent first (as "cloud-native Java application") then talk to YK (or me next Wednesday) about where (if anywhere) it is appropriate to mention Jakarta EE and MicroProfile without it looking crowbarred in.

Author: Done

> Create a new stage in the pipeline code and write all the needed commands between the shell option (sample commands give below):

Reviewer 2: Replace with "In your pipeline code, add the CLI commands in a new stage. The following sample pipeline code does xxxx and xxxx." (explain what will happen if they use the sample code)

Author: Modified to "In your pipeline code, add these CLI commands in a new stage. The following sample pipeline code connects to IBM cloud from CLI and then connects to kubernetes cluster running inside that, further it runs all the kubernetes deployment related configurations ."

>                            sh '''

Reviewer 2: Check the rendering of this tagging. I don't think it's right.

Author: no issues in it. We can discuss further

                              ibmcloud login --apikey $IBM_CLOUD_API_KEY -g $IBM_CLOUD_RSGRP
                              ibmcloud ks cluster config --cluster $CLUSTER-ID
                              kubectl config current-context
                              kubectl create -f deploy/deployment.yaml #( simple k8s deployment command )
                              kubectl create -f deploy/service.yaml #( simple k8s deployment command )
                              kubectl create -f deploy/route.yaml #( simple k8s deployment command )
                           '''

> Make sure that your Kubernetes configuration files are in the same source code repository (under a folder called "deploy"), also the image name should be updated manually or programmatically 

Reviewer 2: Change to "Make sure that your Kubernetes configuration files are stored in the same git repository as your Jenkinsfile in a sub-directory called `deploy`." (end of sentence)

I don't understand the second sentence (from "Also the image name..."). Who should update the image name and why and when?

Author: Modified to "Make sure that your Kubernetes configuration files are stored in the same git repository as your Jenkinsfile in a sub-directory called deploy, also the docker image name in kubernetes deployment configuration file (as shown in screenshot below) should be updated manually or programmatically (in case of any change in it on run time)Make sure that your Kubernetes configuration files are stored in the same git repository as your Jenkinsfile in a sub-directory called deploy, also the docker image name in kubernetes deployment configuration file (as shown in screenshot below) should be updated manually or programmatically (in case of any change in it on run time)

Author: I can explain further when you are back

image::/img/blog/deploymentyaml.png[Image reference in deployment yaml,width=40%,align="center"]

When Jenkins has checked out the Java application code for the code build, all the Kubernetes configuration files are also downloaded to the Jenkins workspace so that Jenkins can run the IBM Cloud and Kubernetes commands to connect to the Kubernetes cluster and deploy the application.

All the other needed application deployment commands are available in this Kubernetes command page that is straightforward 

> link:https://kubernetes.io/docs/reference/kubectl/cheatsheet/[Kubernetes sample commands].

Reviewer 2: Rephrase to "See the Kubernetes documentation for other commands
link:https://kubernetes.io/docs/reference/kubectl/cheatsheet/[Kubernetes sample commands]."

Author: Done

> == QA testing, Jakarta EE & MicroProfile applications with Jenkins

Reviewer 2: "QA testing cloud-native Java applications with Jenkins"

Author: Done

> Apart from running JUnit test cases along with the code build phase, Jenkins can trigger functional and integration QA test cases automatically after deploying the Jakarta EE & MicroProfile application.

Reviewer 2: Change "Jakarta EE & MicroProfile" to "cloud-native Java"

Author: Done

> Configure the test cases on jenkins job and test it manually. Create an "Authentication Token" in "Trigger builds remotely" section under "Build Triggers". Trigger this test case from docker "entrypoint" file by using remote rest API call that uses this authentication token as the identifier

Reviewer 2: Capitalise:
"Jenkins job"
"Docker"

"Create an "Authentication Token" in "Trigger builds remotely" section under "Build Triggers"." - rephrase this so that it either instructs someone to click **Authentication Token** etc, or so that the sentence says "Create an authentication token in the **Trigger builds remotely** section...". Don't mix the names of sections with parts of the sentence.

Author: rephrased as "Create an authentication token in "Trigger builds remotely" section under "Build Triggers"."

> For example: curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<authentication-token>

Reviewer 2: Put the command into command tags (look in another blog post source file or talk to Kin if you need help with tagging).

Author: Done

> Note: Auth token can be generated from postman

Reviewer 2: Change to "You can generate an authentication token with Postman." (and link to Postman?)

Author: modified to "Note: You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the login credentials of Jenkins."

> == Kubernetes monitoring tools

Reviewer 2: "== Monitoring cloud-native Java applications on Kubernetes" ?

Author: No, its better to keep as K8s monitoring tools, because these commands are not restricted to only CN java applications. We can discuss further

> Kubernetes provides commands to check the application or cluster logs and memory/cpu usage through the commands like 

Reviewer 2: "...and memory or CPU usage..."

Reviewer 2: "...commands like:"

Author: modified to "memory and cpu usage"

    $ kubectl logs ..
    $ cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
    $ cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)

For persistence of logs and usage stats, several applications are available in the market, which can be integrated with Kubernetes, details about some of those apps are given below

These tools are deployed in Kubernetes cluster itself where the application is running and exposed by using route and access the gathered details from UI.

    -> https://grafana.com/oss/loki/
    -> https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872
    -> https://prometheus.io
    -> https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/

> 

Reviewer 2: Rephrase to:

"There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. You can read more in the following articles:

* link:https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872[ARTICLE TITLE HERE]
* link:https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/[ARTICLE TITLE HERE]
"

Author: Done

== Conclusion

> There are many ways in which you can configure your DevOps pipeline. This blog post is a quick introduction to how you can use Jenkins to set up a simple CI/CD pipeline to build and deploy your Jakarta EE & MicroProfile applications.

Reviewer 2: change "Jakarta EE & MicroProfile applications." to "cloud-native Java applications on Liberty."

Author: Done

---
layout: post
title: "Continuous integration and delivery (CI/CD) of Jakarta EE & MicroProfile applications with Open Liberty to Kubernetes by using Jenkins"
categories: blog
author_picture: https://avatars3.githubusercontent.com/shamjithantholi
author_github: https://github.com/shamjithantholi
seo-title: TITLE - OpenLiberty.io
seo-description: DESCRIPTION
blog_description: "A high level explanation about how to build a devops pipeline with Jenkins as the core controller component "
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---

> = Continuous integration and delivery (CI/CD) of Jakarta EE & MicroProfile applications with Open Liberty to Kubernetes by using Jenkins

Reviewer 2: The title needs to be shorter but still search engine friendly. Based on YK's comments, how about:
"Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins"
or
"Using Jenkins for continuous integration and delivery (CI/CD) of cloud-native Java applications"

@yeekangc ?

Author: modified title to "Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins"

> Shamjith Antholi https://github.com/shamjithantholi

Reviewer 2: The URL needs triangular brackets round it to render correctly
eg like `Alasdair Nottingham <https://github.com/nottycode>`

Author: Done

:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

> == DevOps, CI/CD, Jenkins & OpenLiberty ( to deploy Jakarta EE & MicroProfile application)

Reviewer 2: Remove the first sub-heading. It's not necessary and not part of our house style. Instead we just need a punchy, search engine-friendly opening paragraph.

Author: removed the first heading

DevOps is a culture or a process, which enables the integration of the software development (dev) and IT operations (ops). It can facilitate the setup of deployment pipeline and helps to improve the speed of software delivery. Automation of the entire build, test, and deployment of the software applications can be achieved through DevOps process. There are tons of efficient applications and process methodologies available in the market that can be readily integrated to the DevOps pipeline, which can help to achieve the wanted results.

> CI/CD are acronym’s for continuous integration, and continuous delivery or deployment used in DevOps. Continuous integration is a process where developers frequently merge code changes into a source code repository, which automatically start code build, security scan, and test execution and deployment. Automation of this process enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process. In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your Jakarta EE & MicroProfile applications 

Reviewer 2: The first two paragraphs are too long and overlap in what they're saying. It needs to be more concise.

Can you try removing the first paragraph completely (which currently starts "DevOps is a culture or process...") then change this second paragraph (which currently starts "CI/CD are acronym's...") to:

"Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then automatically starts a code build, security scanning, and test running, and deployment of the built application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[OpenLiberty]."

Does that make technical sense to the target audience? Happy to tweak but this opening paragraph really needs to be shorter and punchier.

Reviewer 2: Make the words "Open Liberty" in this paragraph a link to https://openliberty.io/

Author: Good .. replaced this para with the first 2 paragraph

> Jenkins is an open source automation server, which helps to automate application code build, test, and deployment. Jenkins comes with a hundred’s of plug-ins, which eases the various DevOps pipeline tool integrations into the DevOps pipeline.

Reviewer 2: Change to something like (give different examples if you like but we're introducing Jenkins here so we should illustrate what we're saying):

"Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. For example, Jenkins can pull code changes from a git repository, build the code using Maven, then run Junit tests against the built code, and deploy the built code to an Open Liberty runtime in Kubernetes."

Author: Done

> Open Liberty is a lightweight Java runtime for building cloud-native applications and microservices. In this blog, it details how you can use Jenkins in CI/CD pipeline for Jakarta EE & MicroProfile application build/test/deployment (to deploy using Openliberty ) . To try Openliberty link:https://openliberty.io[OpenLiberty] 

Reviewer 2: Remove this paragraph about Open Liberty. It's not necessary. I know we've not mentioned Jakarta EE and MicroProfile yet but let's see how the whole post works then we'll look at where it's appropriate to fit those terms in. We can't just stuff terms in - it needs to work for readability too.

Author: Done

The following diagram shows a simple architecture for a pipeline that builds and deploys a Jakarta EE & MicroProfile application in a single Kubernetes environment. When a code change is detected by Jenkins, Jenkins builds the code, runs security and static analysis scans, and runs unit tests against the built code. Jenkins then generates a Docker image of the application with its Open Liberty server and saves the image to a Docker repository, Jenkins then trigger the deployment using CLI commands and deploy the Docker image to Kubernetes..

> image::/img/blog/liberty-devops-generic-architecture-2.png[Liberty devops generic architecture diagram ,width=70%,align="center"]

Reviewer 2: Nice diagram. I can't remember where we got to with checking if we're allowed to use the logos etc. I assume it's okay with @mbroz2 @kinueng ?

Reviewer 2: Change the alt text in that tag to:

"Liberty DevOps architecture diagram,"

Author: I created that diagram. As I know, we can use the logos without any modification in articles about the use of same technology ( verified a few from IBM Colleagues as well )

Reviewer 3: This depends on both the logo/company and how the logo is used (there's a difference between talking about a company in a blog post vs implying a relationship with said company). You have to follow the rules of that company to make sure you're using their logos in a way that complies with their rules. For example, Docker's copyright rules are here: https://www.docker.com/legal/trademark-guidelines/
Kubernetes is under the Linux foundation: https://www.linuxfoundation.org/trademark-usage/

Otherwise, you'd need to seek permission to use their logo (which we've done in the past).  If unsure, I would recommend getting input from @CassTucker

Reviewer 4: Each company has guidelines on how to use their logos. The picture in question here contains many logos that will have different usage guidelines.  

For an example, SonarQube's Logo and Usage https://www.sonarqube.org/logos/ has a easy `Can do` and `Can NOT do`. If the guidelines are met, then the usage of SolarQube is fine per the page saying "You can use the SonarQube logo freely in compliance with the guidelines below."

My recommendation is that we invest the time to ensure we are not violating any logo usage guidelines

> Basic understanding of Git, Docker, Kubernetes are a prerequisite for this blog.

Reviewer 2: Remove this sentence as the next paragraph says exactly the same thing.

Author: Done

open-graph-image: https://openliberty.io/img/blog/liberty-devops-generic-architecture-2.png
---
= Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins
Shamjith Antholi <https://github.com/shamjithantholi>
:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then automatically builds the code, runs security scans, runs tests, and then deploys the application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[Open Liberty].

> Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change in the Git repository is detected by Jenkins, Jenkins builds the code, runs Junit unit tests and then SonarQube security and static analysis scans against the built code. Jenkins then generates a Docker image of the application with its Open Liberty server, scans the image with Trivy, and saves the image to a Docker repository. Jenkins then triggers a deployment using CLI commands and deploys the Docker image to Kubernetes.

Reviewer: ```suggestion
Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change in the Git repository is detected by Jenkins, Jenkins builds the code, runs Junit unit tests and then SonarQube security and static analysis scans against the built code. Jenkins then generates a Docker (container) image of the application with Open Liberty as its runtime, scans the image with Trivy, and saves the image to a Docker repository. Jenkins then triggers a deployment using CLI commands and deploys the Docker image to Kubernetes.
```

> 

Reviewer: Should we consider a reference to GitOps like the below?

"Since Git is our starting point and the CI/CD process is triggered by source changes in Git, this is also known as the GitOps approach to DevOps."

Author: Done.. added "In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[Open Liberty] using GitOps approach".

Reviewer 2: I've tweaked it by adding a shortened version of YK's suggested text as a new sentence after the original sentence. See what you think.

image::/img/blog/liberty-devops-generic-architecture.png[Liberty DevOps architecture diagram,width=70%,align="center"]

In this blog post, I will assume that you have a basic understanding of Git, Docker, Kubernetes. I will also assume that the Jakarta EE and MicroProfile application code is stored in link:https://github.com/[Github]. I will use link:https://hub.docker.com/[DockerHub] to store the Docker images and I will deploy the containerized application to link:https://cloud.ibm.com/kubernetes/catalog/create[IBM Cloud Kubernetes Service].

== Installing and configuring Jenkins to set up CI/CD of a cloud-native Java application

Install link:https://www.jenkins.io/doc/book/installing/[Jenkins] with the following plugins on the base image of your Jenkins container image, or on the Jenkins master.

* Maven, to build java code
* Pipeline, for creating Jenkins pipeline jobs
* Multibranch Scan Webhook Trigger, to create Jenkins pipeline type jobs which pull all branches of code from Git to Jenkins

> * Docker, to build and push container images

Reviewer: ```suggestion
* Docker or equivalent like [Podman](https://podman.io), to build and push container images
```

* Kubernetes, to use the Kubernetes template

> Jenkins builds the Liberty Java code using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].

Reviewer: Is there an alternative to "Jenkins master" that we can use?

Reviewer 2: Ah yes, I think it's "Jenkins controller" now (they've changed "Jenkins slave" to "Jenkins agent"). @shamjithantholi Can you confirm that "Jenkins controller" is what used to be known as "Jenkins master", pls?

> Jenkins builds the Liberty Java code using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].

Reviewer: ```suggestion
Jenkins builds your Java application code running on Liberty using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].
```

== Writing Jenkins pipeline scripts

It is good practice to adhere to the concept of infrastructure as a code (IaaC) when writing DevOps pipelines. Writing Jenkins jobs as pipeline scripts is a good example of IaaC.

You can write link:https://www.jenkins.io/doc/book/pipeline/syntax/[Jenkins pipeline code] in one of the following ways:

> * As pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins master. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 

Reviewer: Is there an alternative to "Jenkins master" that we can use?

Reviewer 2: Yes, it should be "Jenkins controller" instead. @shamjithantholi can you confirm, pls?

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-git.png[Pipeline code stored in a Jenkinsfile on git,width=50%,align="center"]

== Building the cloud-native Java application with Open Liberty on Jenkins

Build your Dockerfile from the foundation code given below (gather the additional files mentioned in this code from Open Liberty support team).

[source]
----
FROM icr.io/appcafe/open-liberty:full-java11-openj9-ubi
# Add Liberty server configuration including all necessary features
COPY --chown=1001:0  server.xml /config/
# Modify feature repository (optional)
COPY --chown=1001:0 featureUtility.properties /opt/ol/wlp/etc/
# This script will add the requested XML snippets to enable Liberty features and grow image to be fit-for-purpose using featureUtility. 
# Only available in 'kernel-slim'. The 'full' tag already includes all features for convenience.
RUN features.sh
# Add interim fixes (optional)
#COPY --chown=1001:0  interim-fixes /opt/ol/fixes/
.
.
.
.
COPY --chown=1001:0 src/main/liberty/config/ /config/
COPY --chown=1001:0 ./target/*.war /config/apps/
RUN configure.sh
----

> The following sample pipeline code builds the Liberty Java code, packages it into a Docker image, and pushes the Docker image to a remote Docker image repository, such as DockerHub:

Reviewer: ```suggestion
The following sample pipeline code builds your Java application code, packages it into a Docker container image, and pushes the Docker image to a remote container image repository, such as Docker Hub or an equivalent within your enterprise:
```

For static code analysis, SonarQube Community edition can be used. SonarQube's Jenkins client setup details are given at link:https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-jenkins/[SonarScanner for Jenkins]. The following command packages the code with Maven and runs the SonarQube scan:

[source]
----
$ mvn package verify sonar:sonar -Dsonar.projectKey=sampleapp -Dsonar.host.url=http://localhost:9000 -Dsonar.login=7b13f240123456780479b13d09ba8
----

For Docker image scanning, you can use link:https://hub.docker.com/r/aquasec/trivy/[Trivy]. This scan provides the vulnerability details of open source JAR files that are used in the application. Sample commands are given below

[source]
----
docker run aquasec/trivy image docker.io/<docker-repo>/liberty-app:v1.0
----

== Deploying cloud-native Java applications with Open Liberty to Kubernetes with Jenkins

For simplicity, I will use the command line (CLI) option to configure Jenkins to deploy a Jakarta EE and MicroProfile application with Open Liberty to Kubernetes. You could use one of many tools though, such as Helm, Travis CI, Circle CI, etc.

> In your pipeline code, add these CLI commands in a new stage. The following sample pipeline code connects to IBM cloud from CLI and then connects to kubernetes cluster running inside that, further it runs all the kubernetes deployment related configurations .

Reviewer: ```suggestion
In your pipeline code, add these CLI commands in a new stage. The following sample pipeline code connects to IBM Cloud from CLI and then connects to Kubernetes cluster running inside that, further it runs all the Kubernetes deployment related configurations .
```

See the Kubernetes documentation for other commands:
link:https://kubernetes.io/docs/reference/kubectl/cheatsheet/[kubectl Cheat Sheet].

== QA testing cloud-native Java applications with Jenkins

Apart from running JUnit test cases along with the code build phase, Jenkins can trigger functional and integration QA test cases automatically after deploying the cloud-native Java application.

Configure the test cases in the Jenkins job and test it manually. Create a remote job identifier authentication token in the "Trigger builds remotely" section under "Build Triggers". Trigger this test case from the Docker "entrypoint" file by using remote rest API call that uses this authentication token as the identifier.

For example, run the following command in a terminal: 

[source]
----
curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<remote-job-identifier-authentication-token>
----

You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the Jenkins login credentials.

== Kubernetes monitoring tools

> Kubernetes provides commands to check the application or cluster logs and memory and cpu usage through the commands like: 

Reviewer: ```suggestion
Kubernetes provides commands to check the application or cluster logs and memory and CPU usage through the commands like: 
```

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-git.png[Pipeline code stored in a Jenkinsfile on git,width=50%,align="center"]

== Building the cloud-native Java application with Open Liberty on Jenkins

Build your Dockerfile from the foundation code given below. Additional files mentioned in the code can be downloaded from link:/extras/featureUtility.properties[featureUtility.properties] , link:/extras/server.xml[server.xml] 

[source]
----
FROM icr.io/appcafe/open-liberty:full-java11-openj9-ubi
# Add Liberty server configuration including all necessary features
COPY --chown=1001:0 server.xml /config/
# Modify feature repository (optional), this file contains the proxy settings to connect to private maven repository

> COPY --chown=1001:0 featureUtility.properties /opt/ol/wlp/etc/

Reviewer 2: If you think it's important to keep these lines in, could we add a sentence after the script saying

"If you are using a private Maven repository, see link:https://openliberty.io/docs/latest/reference/command/featureUtility-commands.html[featureUtility commands]."

Reviewer 2: Alternatively, just remove these lines from the script (and still provide the link to the docs).

* As pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins master. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-jenkins.png[Pipeline code directly on Jenkins,width=70%,align="center"]

* As plain text in a Jenkinsfile (a plain text file) in Git and mapping it to Jenkins. This is better for ensuring that your configuration is always under version control. If you have specific build and deployment configurations for separate environments, such as dev, staging, and production, you create a separate Jenkinsfile for each environment and store it in that environment-specific Git repository branches. You can use either the "Pipeline" or "Multibranch pipeline" type of job in this case.

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-git.png[Pipeline code stored in a Jenkinsfile on git,width=50%,align="center"]

== Building the cloud-native Java application with Open Liberty on Jenkins

> Build your Dockerfile from the foundation code given below (gather the additional files mentioned in this code from Open Liberty support team).

Reviewer 2: "gather the additional files mentioned in this code from Open Liberty support team" - what does this mean? Can't they just get the files from a Liberty installation from Maven? Or do you mean customized files? In this context, can they just use the standard files?

Reviewer: +1 for addressing the above comment.  I believe this has been resolved in the latest PR, right?  Or, no?

Also, I find the references to "foundation code" and "code" odd in this context.  They are more "instructions" or "commands" in a Dockerfile or Containerfile?

For example, run the following command in a terminal: 

[source]
----
curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<remote-job-identifier-authentication-token>
----

You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the Jenkins login credentials.

== Kubernetes monitoring tools

Kubernetes provides commands to check the application or cluster logs and memory and cpu usage through the commands like: 

[source]
----
    $ kubectl logs ..
    $ cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
    $ cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)
----

There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. You can read more in the following articles:

* link:https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872[Grafana with Loki]

* link:https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/[Prometheus setup in Kubernetes]

> These tools are deployed in the Kubernetes cluster where the application is running and exposed by using route and access the gathered details from UI.

Reviewer 2: This sentence doesn't make sense. I think it needs splitting into two and re-wording slightly.

open-graph-image: https://openliberty.io/img/blog/liberty-devops-generic-architecture-2.png
---
= Continuous integration and delivery (CI/CD) of cloud-native Java applications using Jenkins
Shamjith Antholi <https://github.com/shamjithantholi>
:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

Continuous integration and continuous deployment (CI/CD) is a process in which developers frequently merge code changes into a source code repository, which then automatically builds the code, runs security scans, runs tests, and then deploys the application. Automation of this process, through adopting a DevOps culture, tools, and processes, enables organizations to release on a more frequent basis without compromising on quality. There are various tools available to automate the CI/CD process that can be readily integrated into a DevOps deployment pipeline. In this blog post we'll discuss how you can use Jenkins to automate the CI/CD process for building, testing, and deploying your cloud-native Java applications running on link:https://openliberty.io[Open Liberty]. As the CI/CD process is triggered by source changes in Git, this is also known as the GitOps approach to DevOps.

> Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change in the Git repository is detected by Jenkins, Jenkins builds the code, runs Junit unit tests and then SonarQube security and static analysis scans against the built code. Jenkins then generates a Docker (container) image of the application with Open Liberty as its runtime, scans the image with Trivy, and saves the image to a Docker repository. Jenkins then triggers a deployment using CLI commands and deploys the Docker image to Kubernetes.

Reviewer: ```suggestion
Jenkins is a popular open source automation server that has hundreds of plugins to help integrate it into your existing DevOps pipeline. The following diagram shows a simple architecture for a Jenkins deployment pipeline that builds and deploys a Jakarta EE and MicroProfile application in a single Kubernetes environment. When a code change in the Git repository is detected by Jenkins, Jenkins builds the code, runs Junit unit tests and then SonarQube security and static analysis scans against the built code. Jenkins then generates a Docker container image of the application with Open Liberty as its runtime, scans the image with Trivy, and saves the image to a Docker repository. Jenkins then triggers a deployment using CLI commands and deploys the Docker image to Kubernetes.
```

Author: Done

image::/img/blog/liberty-devops-generic-architecture.png[Liberty DevOps architecture diagram,width=70%,align="center"]

* *Jenkins* link:https://www.jenkins.io[https://www.jenkins.io]
* *SonarQube* link:https://www.sonarqube.org[https://www.sonarqube.org]
* *Aqua Trivy* link:https://www.aquasec.com[https://www.aquasec.com]
* *Docker* link:https://www.docker.com[https://www.docker.com]
* *Kubernetes* link:https://kubernetes.io[https://kubernetes.io]

> In this blog post, I will assume that you have a basic understanding of Git, Docker, Kubernetes. I will also assume that the Jakarta EE and MicroProfile application code is stored in link:https://github.com/[Github]. I will use link:https://hub.docker.com/[DockerHub] to store the Docker images and I will deploy the containerized application to link:https://cloud.ibm.com/kubernetes/catalog/create[IBM Cloud Kubernetes Service].

Reviewer: ```suggestion
In this blog post, I will assume that you have a basic understanding of Git, Docker, Kubernetes. I will also assume that the Jakarta EE and MicroProfile application code is stored in link:https://github.com/[GitHub]. I will use link:https://hub.docker.com/[Docker Hub] to store the Docker container images and I will deploy the containerized application to link:https://cloud.ibm.com/kubernetes/catalog/create[IBM Cloud Kubernetes Service].
```

Author: Done

== Installing and configuring Jenkins to set up CI/CD of a cloud-native Java application

> Install link:https://www.jenkins.io/doc/book/installing/[Jenkins] with the following plugins on the base image of your Jenkins container image, or on the Jenkins master.

Reviewer: ```suggestion
Install link:https://www.jenkins.io/doc/book/installing/[Jenkins] with the following plugins on the base image of your Jenkins container image, or on the Jenkins controller.
```

* Maven, to build java code
* Pipeline, for creating Jenkins pipeline jobs
* Multibranch Scan Webhook Trigger, to create Jenkins pipeline type jobs which pull all branches of code from Git to Jenkins
* Docker or equivalent like link:https://podman.io[Podman], to build and push container images
* Kubernetes, to use the Kubernetes template

> Jenkins builds your Java application code running on Liberty using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins master) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].

Reviewer: ```suggestion
Jenkins builds your Java application code running on Liberty using Jenkins pipeline scripts. The script can run directly on your Jenkins installation (known as Jenkins controller) but, if the application is big, you need to use a Jenkins agent. For more information about about setting up Jenkins agents, see link:https://www.jenkins.io/doc/book/using/using-agents/[Using Jenkins agents] and link:[https://www.jenkins.io/doc/book/pipeline/syntax/Pipeline Syntax].
```

Author: Done

== Writing Jenkins pipeline scripts

> It is good practice to adhere to the concept of infrastructure as a code (IaaC) when writing DevOps pipelines. Writing Jenkins jobs as pipeline scripts is a good example of IaaC.

Reviewer: ```suggestion
It is good practice to adhere to the concept of Infrastructure as Code (IaC) when writing DevOps pipelines. Writing Jenkins jobs as pipeline scripts is a good example of IaC.
```

Author: Done

You can write link:https://www.jenkins.io/doc/book/pipeline/syntax/[Jenkins pipeline code] in one of the following ways:

> * As pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins master. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 

Reviewer: ```suggestion
* As pipeline code written directly in the link:https://www.jenkins.io/doc/book/pipeline/getting-started/[Jenkins UI] and stored in the Jenkins controller. This is a useful way to get started. You would need to take a backup of the Jenkins instance to store the pipeline code because it is part of the Jenkins installation. 
```

Author: Done

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-jenkins.png[Pipeline code directly on Jenkins,width=70%,align="center"]

* As plain text in a Jenkinsfile (a plain text file) in Git and mapping it to Jenkins. This is better for ensuring that your configuration is always under version control. If you have specific build and deployment configurations for separate environments, such as dev, staging, and production, you create a separate Jenkinsfile for each environment and store it in that environment-specific Git repository branches. You can use either the "Pipeline" or "Multibranch pipeline" type of job in this case.

[.imageblock.img_border_light]
image::/img/blog/pipeline-code-on-git.png[Pipeline code stored in a Jenkinsfile on git,width=50%,align="center"]

== Building the cloud-native Java application with Open Liberty on Jenkins

> Build your Dockerfile from the following code:

Reviewer: ```suggestion
Build the container image for your application running on Liberty using a Dockerfile with the following content:
```

Author: Done

[source]
----
FROM icr.io/appcafe/open-liberty:full-java11-openj9-ubi
# Add Liberty server configuration including all necessary features
COPY --chown=1001:0 server.xml /config/
# Modify feature repository (optional), this file contains the proxy settings to connect to private maven repository
COPY --chown=1001:0 featureUtility.properties /opt/ol/wlp/etc/
# This script will add the requested XML snippets to enable Liberty features and grow image to be fit-for-purpose using featureUtility. 
# Only available in 'kernel-slim'. The 'full' tag already includes all features for convenience.
RUN features.sh
# Add interim fixes (optional)
#COPY --chown=1001:0  interim-fixes /opt/ol/fixes/
.
.
.
.
COPY --chown=1001:0 src/main/liberty/config/ /config/
COPY --chown=1001:0 ./target/*.war /config/apps/
RUN configure.sh
----

> If you include the optional `featureUtility.properties` line, see link:/docs/latest/reference/command/featureUtility-commands.html#_define_custom_repositories[featureUtility commands] in the Open Liberty docs for how to create the properties file.

Reviewer: ```suggestion
If you include the optional `featureUtility.properties` line, see link:/docs/latest/reference/command/featureUtility-commands.html#_define_custom_repositories[featureUtility commands] in the Open Liberty docs for how to create the properties file.

You can learn more about the Liberty container images available to you and best practices for creating a Dockerfile link:https://github.com/OpenLiberty/ci.docker/[here].
```

Author: Done

> The following sample pipeline code builds your Java application code, packages it into a Docker container image, and pushes the Docker image to a remote container image repository, such as Docker Hub or an equivalent within your enterprise:

Reviewer: ```suggestion
The following sample pipeline code builds your Java application code, packages it into a Docker container image, and pushes the container image to a remote container image repository, such as Docker Hub or an equivalent within your enterprise:
```

Author: Done

                         docker push <remote-docker-image-repository-url>/<docker-repo-name>/liberty-$<code identifier>:$<docker image version>
                         #eg: docker push docker.io/someid/liberty-app:v1.0  
                       '''   
                  }
               }
            }
          }
----

Where:

* _git-token_ is the personal access token you have generated from your Github account.
* _remote-docker-image-repository-url_ is the location of the Docker image repository.
* _username_ is your user name for the Docker image repository.
* _password_ is your password for the Docker image repository.
* _docker-image-version_ is the version number of your Docker image, a unique identifier.

For static code analysis, you can use SonarQube Community edition. SonarQube's Jenkins client setup details are given at link:https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-jenkins/[SonarScanner for Jenkins]. The following Maven command packages the code with Maven and runs the SonarQube scan:

[source]
----
mvn package verify sonar:sonar -Dsonar.projectKey=sampleapp -Dsonar.host.url=http://localhost:9000 -Dsonar.login=7b13f240123456780479b13d09ba8
----

> For Docker image scanning, you can use link:https://hub.docker.com/r/aquasec/trivy/[Trivy]. This scan provides the vulnerability details of open source JAR files that are used in the application. The following Docker command runs Trivy against your Docker image:

Reviewer: ```suggestion
For container image scanning for security purpose, you can use link:https://hub.docker.com/r/aquasec/trivy/[Trivy]. This scan provides the vulnerability details of open source JAR files that are used in the application. The following Docker command runs Trivy against your container image:
```

Author: Done

For example, run the following command in a terminal: 

[source]
----
curl -I -u <auth-token> https://<jenkins-host>/job/<job-name>/build?token=<remote-job-identifier-authentication-token>
----

You can generate an authentication token (auth-token) with link:https://www.postman.com[Postman] using the Jenkins login credentials.

== Kubernetes monitoring tools

Kubernetes provides commands to check the application or cluster logs and memory and CPU usage through the commands like: 

[source]
----
kubectl logs ..
cat /sys/fs/cgroup/cpu/cpuacct.usage (after connecting to k8s pod)
cat /sys/fs/cgroup/memory/memory.usage_in_bytes (after connecting to k8s pod)
----

There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. You can read more in the following articles:

* link:https://medium.com/nerd-for-tech/logging-at-scale-in-kubernetes-using-grafana-loki-3bb2eb0c0872[Grafana with Loki]

* link:https://k21academy.com/docker-kubernetes/prometheus-grafana-monitoring/[Prometheus setup in Kubernetes]

> 

Reviewer: Consider link(s) to Liberty-related assets like https://community.ibm.com/community/user/wasdevops/blogs/don-bourne1/2021/06/26/metrics-and-monitoring-guidance-for-open-liberty-a?

We can consult @donbourne for what is best to reference too.

Thanks.

Author: Removed all except "There are several applications available that you can integrate with Kubernetes to persist logs and usage statistics, such as link:https://grafana.com/oss/loki/[Grafana] and link:https://prometheus.io[Prometheus]. "
